install/first_halyard_boot.sh

```diff
@@ -0,0 +1,202 @@
+  # Keep the original around for reference.
+  # From now on, all we need to do is start_spinnaker
+  local original=$(get_instance_metadata_attribute "startup-script")
+  echo "$original" > "$SPINNAKER_INSTALL_DIR/scripts/original_startup_script.sh"
+  clear_instance_metadata "startup-script"
+}
+
+function install_halyard() {
+  wget https://raw.githubusercontent.com/spinnaker/halyard/master/InstallHalyard.sh
+  bash InstallHalyard.sh -y
+  rm InstallHalyard.sh
+
+  local version=$(get_instance_metadata_attribute "version")
+  hal config version edit --version $version
+}
+
+function configure_docker() {
+  local config_path="$GCR_DIR/account.json"
+  mkdir -p $(dirname $config_path)
+  chown -R spinnaker:spinnaker $(dirname $config_path)
+  
+  if clear_metadata_to_file "gcr_cred" $config_path; then
+    if [[ -s $config_path ]]; then
+      chmod 400 $config_path
+      chown spinnaker $config_path
+      echo "Successfully wrote gcr account to $config_path"
+    else
+      echo "Failed to write gcr account to $config_path"
+      rm $config_path
+
+      return 1
+    fi
+  fi
+
+  hal config provider docker-registry enable
+  hal config proivder docker-registry account add $DOCKER_ACCOUNT \
+    --password-file $config_path \
+    --username _json_key \
+    --address gcr.io
+}
+
+function configure_kubernetes() {
+  local config_path="$KUBE_DIR/config"
+  mkdir -p $(dirname $config_path)
+  chown -R spinnaker:spinnaker $(dirname $config_path)
+  
+  if clear_metadata_to_file "kube_config" $config_path; then
+    if [[ -s $config_path ]]; then
+      chmod 400 $config_path
+      chown spinnaker $config_path
+      echo "Successfully wrote kube_config to $config_path"
+    else
+      echo "Failed to write kube_config to $config_path"
+      rm $config_path
+
+      return 1
+    fi
+  fi
+
+  hal config provider kubernetes enable
+  hal config provider kubernetes account add $KUBERNETES_ACCOUNT \
+    --kubeconfig-path $config_path \
+    --docker-registries $DOCKER_ACCOUNT
+}
+
+function configure_google() {
+  local config_path="$GCE_DIR/account.json"
+  mkdir -p $(dirname $config_path)
+  chown -R spinnaker:spinnaker $(dirname $config_path)
+  local project=$(get_instance_metadata_attribute "project")
+  
+  if clear_metadata_to_file "gce_cred" $config_path; then
+    if [[ -s $config_path ]]; then
+      chmod 400 $config_path
+      chown spinnaker $config_path
+      echo "Successfully wrote gce credential to $config_path"
+    else
+      echo "Failed to write gce credential to $config_path"
+      rm $config_path
+
+      return 1
+    fi
+  fi
+
+  hal config provider google enable
+  hal config provider google account add  $GOOGLE_ACCOUNT \
+    --json-path $config_path \
+    --project $project
+}
+
+function configure_storage() {
+  local bucket_name=$(get_instance_metadata_attribute "bucket_name")
+  hal config storage edit --account-name $GOOGLE_ACCOUNT --bucket $bucket_name
+}
+
+function configure_jenkins() {
```

Comment:
This is starting to feel too tedious.
I recognize at least some of the value of individual attributes, but it means that this script is really brittle. I think you want to pass a configuration file through as metadata and then let hal interpret that. Doing so would eliminate the brittleness of the stack between the person specifying the configuration and hal in the deployment. Otherwise you are passing all these properties through everywhere which is a complicated and tedious interface.

For example, you are going to want to configure monitoring (and that could be one of a few different systems). There are probably echo notifications you want to configure too.

The file you pass through might be two -- a control and data. Or just one. One should be a tar file so you can attach different data files (credentials). And for that matter, maybe you can eventually encrypt it so as not to expose the passwords and credentials through the metadata and other parts of the stack between the creator and hal runtime.

In itself there is nothing here that isnt common practice on GCE (e.g. bitnami jenkins installation) but I dont think they scale to the diversity of Spinnaker. And there's too much independent evolution going on for this level of brittleness.